homeassistant/components/sensor/__init__.py

```diff
@@ -510,65 +567,91 @@ def state(self) -> Any:
 
         # If the sensor has neither a device class, a state class nor
         # a unit_of measurement then there are no further checks or conversions
-        if not device_class and not state_class and not unit_of_measurement:
+        if (
+            not device_class
+            and not state_class
+            and not unit_of_measurement
+            and (prec := self.precision) is None
+        ):
             return value
 
-        if not self._invalid_numeric_value_reported and not isinstance(
-            value, (int, float, Decimal)
-        ):
+        numerical_value: int | float | Decimal | None = None
+        numerical_conversion_failed = None
+        if not isinstance(value, (int, float, Decimal)):
             try:
-                _ = float(value)  # type: ignore[arg-type]
+                if isinstance(value, str) and "." not in value:
+                    numerical_value = int(value)
+                else:
+                    numerical_value = float(value)  # type:ignore[arg-type]
             except (TypeError, ValueError):
-                # This should raise in Home Assistant Core 2023.4
-                self._invalid_numeric_value_reported = True
-                report_issue = self._suggest_report_issue()
-                _LOGGER.warning(
-                    "Sensor %s has device class %s, state class %s and unit %s "
-                    "thus indicating it has a numeric value; however, it has the "
-                    "non-numeric value: %s (%s); Please update your configuration "
-                    "if your entity is manually configured, otherwise %s",
-                    self.entity_id,
-                    device_class,
-                    state_class,
-                    unit_of_measurement,
-                    value,
-                    type(value),
-                    report_issue,
-                )
-                return value
+                # Raise if precision is not None, for other cases log a warning
+                if prec is not None:
+                    raise
+                numerical_conversion_failed = True
```

Comment:
I think we can move the warning back in here and drop the numerical_conversion_failed variable